File: backend/app/schemas/entity_profile.py
"""Pydantic schemas for entity profile endpoints."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class EntityType(str, Enum):
    author = "author"
    publisher = "publisher"
    binder = "binder"


class BiographicalFact(BaseModel):
    """A single biographical fact / gossip item."""

    text: str = Field(..., description="The story (1-2 sentences)")
    year: int | None = Field(None, description="Year the event occurred")
    significance: str = Field(..., description="revelation, notable, or context")
    tone: str = Field(..., description="dramatic, scandalous, tragic, intellectual, or triumphant")
    display_in: list[str] = Field(
        default_factory=list, description="hero-bio, timeline, hover-tooltip"
    )


class RelationshipNarrative(BaseModel):
    """Full relationship story for a high-impact connection."""

    summary: str = Field(..., description="One-line summary for card display")
    details: list[BiographicalFact] = Field(default_factory=list, description="Story facts")
    narrative_style: str = Field(
        "prose-paragraph",
        description="prose-paragraph, bullet-facts, or timeline-events",
    )


class ProfileEntity(BaseModel):
    """Entity summary in profile response."""

    id: int
    type: EntityType
    name: str
    birth_year: int | None = None
    death_year: int | None = None
    founded_year: int | None = None
    closed_year: int | None = None
    era: str | None = None
    tier: str | None = None


class ProfileData(BaseModel):
    """AI-generated profile content."""

    bio_summary: str | None = None
    personal_stories: list[BiographicalFact] = Field(default_factory=list)
    is_stale: bool = False
    generated_at: datetime | None = None
    model_version: str | None = None


class ProfileBook(BaseModel):
    """A book in the entity's collection."""

    id: int
    title: str
    year: int | None = None
    condition: str | None = None
    edition: str | None = None


class ProfileConnection(BaseModel):
    """A connection to another entity."""

    entity: ProfileEntity
    connection_type: str
    strength: int
    shared_book_count: int
    shared_books: list[ProfileBook] = Field(default_factory=list)
    narrative: str | None = None
    narrative_trigger: str | None = None
    is_key: bool = False
    relationship_story: RelationshipNarrative | None = None


class ProfileStats(BaseModel):
    """Collection statistics for this entity."""

    total_books: int = 0
    total_estimated_value: float | None = None
    first_editions: int = 0
    date_range: list[int] = Field(default_factory=list)


class EntityProfileResponse(BaseModel):
    """Full entity profile response."""

    entity: ProfileEntity
    profile: ProfileData
    connections: list[ProfileConnection] = Field(default_factory=list)
    books: list[ProfileBook] = Field(default_factory=list)
    stats: ProfileStats

File: frontend/src/types/entityProfile.ts
/**
 * Entity Profile types.
 * Maps to backend EntityProfileResponse schema.
 */

import type { NodeType, Era, Tier } from "@/types/socialCircles";

// === Gossip Types ===

export type Significance = "revelation" | "notable" | "context";
export type Tone = "dramatic" | "scandalous" | "tragic" | "intellectual" | "triumphant";
export type DisplayLocation = "hero-bio" | "timeline" | "hover-tooltip" | "connection-detail";
export type NarrativeStyle = "prose-paragraph" | "bullet-facts" | "timeline-events";
export type NarrativeTrigger =
  | "cross_era_bridge"
  | "social_circle"
  | "hub_figure"
  | "influence_chain"
  | null;

export interface BiographicalFact {
  text: string;
  year?: number;
  significance: Significance;
  tone: Tone;
  display_in: DisplayLocation[];
}

export interface RelationshipNarrative {
  summary: string;
  details: BiographicalFact[];
  narrative_style: NarrativeStyle;
}

// === Profile Response Types ===

export interface ProfileEntity {
  id: number;
  type: NodeType;
  name: string;
  birth_year?: number;
  death_year?: number;
  founded_year?: number;
  closed_year?: number;
  era?: Era;
  tier?: Tier;
}

export interface ProfileData {
  bio_summary: string | null;
  personal_stories: BiographicalFact[];
  is_stale: boolean;
  generated_at: string | null;
  model_version: string | null;
}

export interface ProfileConnection {
  entity: ProfileEntity;
  connection_type: string;
  strength: number;
  shared_book_count: number;
  shared_books: ProfileBook[];
  narrative: string | null;
  narrative_trigger: NarrativeTrigger;
  is_key: boolean;
  relationship_story: RelationshipNarrative | null;
}

export interface ProfileBook {
  id: number;
  title: string;
  year?: number;
  condition?: string;
  edition?: string;
}

export interface ProfileStats {
  total_books: number;
  total_estimated_value: number | null;
  first_editions: number;
  date_range: number[];
}

export interface EntityProfileResponse {
  entity: ProfileEntity;
  profile: ProfileData;
  connections: ProfileConnection[];
  books: ProfileBook[];
  stats: ProfileStats;
}
